cmd/iam.go

```diff
@@ -1425,6 +1425,37 @@ func (sys *IAMSys) GetServiceAccount(ctx context.Context, accessKey string) (aut
 	return sa, embeddedPolicy, nil
 }
 
+// GetLDAPUserForSvcAcc - gets the LDAP short username who owns the given
+// service account. Returns an empty username if the account is not owned by an
+// LDAP user.
+func (sys *IAMSys) GetLDAPUserForSvcAcc(ctx context.Context, accessKey string) (string, error) {
+	if !sys.Initialized() {
+		return "", errServerNotInitialized
+	}
+
+	if sys.usersSysType != LDAPUsersSysType {
+		return "", nil
+	}
+
+	sys.store.rlock()
+	defer sys.store.runlock()
+
+	sa, ok := sys.iamUsersMap[accessKey]
+	if !ok || !sa.IsServiceAccount() {
+		return "", errNoSuchServiceAccount
+	}
+
+	jwtClaims, err := auth.ExtractClaims(sa.SessionToken, globalActiveCred.SecretKey)
+	if err == nil {
+		lu, ok := jwtClaims.Lookup(ldapUserN)
+		if ok {
+			return lu, nil
+		}
+	}
+
+	return "", nil
```

Comment:
I believe that would make error handling a bit harder (introducing a new error). Currently this code is used like:

```
	u, err := globalIAMSys.GetLDAPUserForSvcAcc(ctx, accessKey)
	if err != nil {
		writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
		return
	}
	if u != "" {
		err = globalClusterReplMgr.IAMChangeHook(ctx, madmin.CRIAMItem{
<SNIP/>
```

If we returned an error when the user is not in LDAP, it would look like:

```
	u, err := globalIAMSys.GetLDAPUserForSvcAcc(ctx, accessKey)
	if err != nil && err != errUserNotInLDAP {
		writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
		return
	}
	if err == nil {
		// USE **u**
		err = globalClusterReplMgr.IAMChangeHook(ctx, madmin.CRIAMItem{
<SNIP/>

```

The function has a comment to say that it returns empty if the user is not in LDAP - this should be ok?

Reply:
The second code is better, because the function name is called GetLDAPUserForSvcAcc. I am more inclined in making functions do the intuitive thing which is returning an error when the ldap user doesn't exist.

Reply:
I've changed the function to do something else now.